Allow DictReader to accept a list of lines of input.

lucaswiman opened the pull request:
# Changes
Fixes #1, wherein `csv.DictReader` can read from a list of lines, but `csv342.DictReader` cannot. The fix here was to construct an `io.StringIO` instance containing the equivalent CSV text.

I also fixed an unrelated test failure, where the test asserted that StringIO.StringIO instances would not be accepted. This was failing for two reasons:
1. `StringIO` was being used as a context manager, which is not supported. This led to an `AttributeError`.
2. The code actually did accept `StringIO.StringIO` instances, which is problematic from the perspective of 2/3 compatibility. `StringIO.StringIO` switches types depending on what data is put into it, and so cannot provide an equivalent interface to Python 3. It also makes sense to force users of this library to only use the python-3-compatible `io.StringIO` buffer.

# Testing

I added a test case reproducing the failure described in #1, and verified that it fails. It passes after 6f4e79cfa081136e5308229df0232cef18f7fd5d. I also verified that the whole test suite now passes on Python 2.

# Thanks!

This library looks to be a great way of managing the upgrade to Python 3. Inconsistencies of the CSV packages between the two versions had made writing compatible code very difficult in an application I'm working on, so `csv342` is a great fit. Thanks for writing this library!

lucaswiman committed:
Add failing test reproducing #1.

lucaswiman committed:
Accept lists of lines.

lucaswiman committed:
Fix test that was failing.

In python 2, StringIO.StringIO is not a context manager, so the test was
failing with an AttributeError. On fixing that problem, the test was still
failing since csv342.reader actually did accept StringIO.StringIO instances.

roskakori commented on csv342.py:
```diff
@@ -127,6 +127,8 @@ class _Utf8Recoder(object):
         """
         def __init__(self, text_stream):
+            if isinstance(text_stream, StringIO.StringIO):
+                raise Error('Cannot read from StringIO.StringIO instances.')
```
If I understand your comment correctly, reading from `io.StringIO` would work. If so, the error message should state this (instead of stating was does not work):
```python
raise Error('StringIO.StringIO must be changed to io.StringIO in order to read a CSV from it')

roskakori commented on test/test_csv342.py:
```diff
@@ -95,8 +95,8 @@ def test_can_read_from_StringIO(self):
-            with StringIO.StringIO('a') as csv_stream:
-                self.assertRaises(csv.Error, csv.reader, csv_stream)
+            csv_stream = StringIO.StringIO('a')
+            self.assertRaises(csv.Error, csv.reader, csv_stream)
```
This causes a minor resource leak. Wrap it in a `try: ... finally: csv_stream.close()`.

lucaswiman commented:
It seems `csv.reader` also accepts a list of lines of input, so I'll need to make some additional changes.

lucaswiman committed:
Use an iterator on the input stream.

The builtin csv.reader and csv.DictReader in both Python 2 and Python 3
seem to consider their input as an iterable of lines. This works for
input streams (e.g. StringIO), but also allows passing in a list of
lines. This commit updates the _Utf8Recoder class to function similarly.

lucaswiman commented:
OK, I updated this to a more general fix that works for both `reader` and `DictReader`. The tests still pass on both Python 2.7 and Python 3.6.

nickpell commented:
Are there plans to make a new pypi release with this change included?

roskakori commented:
@nickpell Version 1.0.0 is now available from PyPI. Thanks for the reminder.
